### PR TITLE
Adding login functionality

### PR DESCRIPTION
**Pull request by rudy-patel**
### What is it?
This PR adds **login functionality** to the app! Now, you can login as a user, register as a new user, and log out. You log in with your **email and password**. This PR also fixes the UI test toast issue (fingers crossed)!

After this PR is merged, I suggest that **everybody creates their own user**, so you have your own list of habits/events to work with, independent from other members. Following this, all **UI tests** run on their own **test user**, so it won't mess up any ongoing work :)

**Note**: Although the number of files changed may look daunting, about **half** of the files changed are just updating our queries to use the user's **ID** instead of the user's **username**.

### Tasks
- [x] run unit tests
- [x] run UI tests
- [x] comments + documentation

### Screenshots
* Login screen:
image

* Sign up screen:
image

Related to #63.

### CODE REVIEW

File: app/build.gradle
Comment (rudy-patel) [inline app/build.gradle:64]: Dependency needed to user Firebase Authorization functionality

File: app/src/main/java/com/example/habitsmasher/UserValidator.java
Comment (jaskim9824) [inline app/src/main/java/com/example/habitsmasher/UserValidator.java:23]: Just wondering, why are the EditText passed into the validator instead of the string value contained within them?

File: app/src/main/java/com/example/habitsmasher/UserValidator.java
Comment (jaskim9824) [inline app/src/main/java/com/example/habitsmasher/UserValidator.java:85]: What exactly are the purpose of these methods? They seem to be used only once within the code (not including unit tests) and they only are used to pass a valid field into a method.

File: app/src/main/java/com/example/habitsmasher/UserValidator.java
Comment (jaskim9824) [inline app/src/main/java/com/example/habitsmasher/UserValidator.java:104]: What was the rationale for splitting up the isVaild into two methods here?

File: app/src/main/java/com/example/habitsmasher/UserValidator.java
Comment (jaskim9824) [inline app/src/main/java/com/example/habitsmasher/UserValidator.java:150]: Minor: we should extract 6 and 15 into constants.

**Review by jaskim9824 — COMMENTED** [review]
Left some questions about the UserVaildator class that I would like answered before I give this my approval

**Comment by rudy-patel** [comment]
> Left some questions about the UserVaildator class that I would like answered before I give this my approval

@jaskim9824 Replied to all of your comments!

**Comment by jaskim9824** [comment]
@rudy-patel I think the replies might have not been saved because I can't see them, or GitHub is acting up for me

File: app/src/main/java/com/example/habitsmasher/UserValidator.java
Comment (JacobNguyen71) [inline app/src/main/java/com/example/habitsmasher/UserValidator.java:103]: Just wondering, so this line(103) checks whether an email is valid? If so, does it check for emails that are not just Gmails? For example, would Yahoo and Microsoft emails work? Another thing, would emails that end in different top level domains be accepted as well?

**Comment by rudy-patel** [comment]
> @rudy-patel I think the replies might have not been saved because I can't see them, or GitHub is acting up for me

@jaskim9824 Just need to scroll up a bit in the **conversation** tab. They are not there anymore because the file got updated.

File: app/src/main/java/com/example/habitsmasher/UserValidator.java
Comment (rudy-patel) [reply app/src/main/java/com/example/habitsmasher/UserValidator.java:103]: Good question! Just ran some tests, and it looks like all email strings are accepted. Yahoo, Microsoft, Outlook, anything. As long as it has an **at sign (@)** and ends in some sort of domain name ending (**..com, .ca, etc.)**

File: app/src/main/java/com/example/habitsmasher/LoginFragment.java
Comment (Nova-Drive) [inline app/src/main/java/com/example/habitsmasher/LoginFragment.java:91]: Why initialize the `loginButton` in the main function and pass it as a parameter, when you've initialized the `registerButton` in the `setClickListenerForRegisterButton` function?

File: app/src/main/java/com/example/habitsmasher/UserRegistrationFragment.java
Comment (Nova-Drive) [inline app/src/main/java/com/example/habitsmasher/UserRegistrationFragment.java:77]: Again, why not initialize the button inside of the function and not pass it as a parameter?

File: app/src/main/java/com/example/habitsmasher/LoginFragment.java
Comment (rudy-patel) [reply app/src/main/java/com/example/habitsmasher/LoginFragment.java:92]: Something I overlooked when extracting these methods... I can move the register button initialization to the main function instead to match this method. 

Good catch. Fixed!

**Comment by jaskim9824** [comment]
@rudy-patel Weird, because on my end my comments aren't deleted. I have no idea why,

image

File: app/src/main/java/com/example/habitsmasher/UserValidator.java
Comment (jaskim9824) [reply app/src/main/java/com/example/habitsmasher/UserValidator.java:25]: I see. Thanks for explaining!

File: app/src/main/java/com/example/habitsmasher/UserRegistrationFragment.java
Comment (Nova-Drive) [inline app/src/main/java/com/example/habitsmasher/UserRegistrationFragment.java:167]: Snazzy, I like this.

File: app/src/main/java/com/example/habitsmasher/UserValidator.java
Comment (jaskim9824) [reply app/src/main/java/com/example/habitsmasher/UserValidator.java]: I was more asking about why there are two variants of each isVaild method. There are two isNewUserValid methods and two isLoginValid methods.

File: app/src/main/java/com/example/habitsmasher/UserRegistrationFragment.java
Comment (Nova-Drive) [inline app/src/main/java/com/example/habitsmasher/UserRegistrationFragment.java:181]: Same thing as other comments, Button in parameters

File: app/src/main/java/com/example/habitsmasher/UserValidator.java
Comment (Nova-Drive) [reply app/src/main/java/com/example/habitsmasher/UserValidator.java:25]: I think I get it, he uses them to set the error messages when not properly formatted. I believe this is in response to the issues with toast messages and detecting them.

**Review by JacobNguyen71 — APPROVED** [review]
Looks good to me, nice work!

File: app/src/main/java/com/example/habitsmasher/UserValidator.java
Comment (Nova-Drive) [inline app/src/main/java/com/example/habitsmasher/UserValidator.java:112]: I can sorta guess what `setError` does, but what does `requestFocus` do? I remember seeing a bunch of StackOverflow people talking about focus, but I never really understood it.

File: app/src/main/java/com/example/habitsmasher/UserValidator.java
Comment (jaskim9824) [reply app/src/main/java/com/example/habitsmasher/UserValidator.java:87]: I see. Though my concern is that these methods are only called in sections when it is certain that a user/login is valid,  so an additional check when retrieving the strings is redundant.

File: app/src/main/java/com/example/habitsmasher/ui/dashboard/HabitListFragment.java
Comment (Nova-Drive) [inline app/src/main/java/com/example/habitsmasher/ui/dashboard/HabitListFragment.java:279]: Could this be taken out and used as a general function? I feel this can be extracted.

File: app/src/main/java/com/example/habitsmasher/UserValidator.java
Comment (rudy-patel) [reply app/src/main/java/com/example/habitsmasher/UserValidator.java:87]: I see what you're saying, and it holds true for **this** stage of the project. But, in the future, if someone wants to go in and use the validator **without** properly validating beforehand, fetching these values from a "validator" implies that the data is valid.

I don't think it hurts so much to check once again to be sure before returning the value to the caller.

File: app/src/main/java/com/example/habitsmasher/ui/profile/ProfileFragment.java
Comment (Nova-Drive) [inline app/src/main/java/com/example/habitsmasher/ui/profile/ProfileFragment.java:54]: Button in parameters

File: app/src/main/java/com/example/habitsmasher/UserValidator.java
Comment (rudy-patel) [reply app/src/main/java/com/example/habitsmasher/UserValidator.java]: I see. I'm just **overriding** these methods to use internally. You can see they are `private`. They are used in the getters/setters to keep those methods short (since we already have access to username, email, and password at the class level, no need to pass it in). 

These methods with no parameters end up calling the main methods anyways. This is a common design pattern.

**Review by Nova-Drive — COMMENTED** [review]
This is really nice! Only gripes is with the buttons passed as parameters, but aside from that you did a really good job.

File: app/src/main/java/com/example/habitsmasher/UserValidator.java
Comment (rudy-patel) [reply app/src/main/java/com/example/habitsmasher/UserValidator.java:112]: Good question, `requestFocus` sets the cursor to the specified field. So in this case, we want the user to **fix** the error, so this conveniently brings the cursor to the incorrect field.

**Review by jaskim9824 — APPROVED** [review]

File: app/src/main/java/com/example/habitsmasher/UserRegistrationFragment.java
Comment (rudy-patel) [reply app/src/main/java/com/example/habitsmasher/UserRegistrationFragment.java:181]: I think we can just pick **one** way and stick with it. Right now, I'm just choosing to **initialize** beforehand and pass the reference to the button to set the listener.

This should be fine.

The other way you're suggesting is equally correct, but it's not the biggest deal, to be honest, and I don't want to lose the current approvals.

File: app/src/main/java/com/example/habitsmasher/UserRegistrationFragment.java
Comment (rudy-patel) [reply app/src/main/java/com/example/habitsmasher/UserRegistrationFragment.java:77]: I think we can just pick **one** way and stick with it. Right now, I'm just choosing to **initialize** beforehand and pass the reference to the button to set the listener.

This should be fine.

The other way you're suggesting is equally correct, but it's not the biggest deal, to be honest, and I don't want to lose the current approvals.

**Pull request merged**
